src/combinatorics/simple_graph/adj_matrix.lean

```diff
@@ -15,89 +16,247 @@ properties to computational properties of the matrix.
+  matrix V V α :=
+λ i j, ite (i = j) 0 (ite (A i j = 0) 1 0)
+
+section compl
+
+variables [decidable_eq α] [decidable_eq V] (A : matrix V V α)
+
+@[simp]
+lemma compl_apply_diag [has_zero α] [has_one α] (i : V) :
+  A.compl i i = 0 :=
+by simp [compl]
+
+@[simp]
+lemma compl_apply [has_zero α] [has_one α] (i j : V) :
+  A.compl i j = 0 ∨ A.compl i j = 1 :=
+begin
+  unfold compl,
+  by_cases h : i = j, {simp [h]},
+  by_cases g : A i j = 0; simp [h , g]
+end
+
+@[simp]
+lemma is_symm_compl [has_zero α] [has_one α] (h : A.is_symm) :
+  A.compl.is_symm :=
+begin
+  ext,
+  unfold compl,
+  congr' 1, {tidy},
+  congr' 1,
+  apply propext,
+  rw [h.apply i j]
+end
+
+@[simps]
+lemma is_adj_matrix_compl [has_zero α] [has_one α] (h : A.is_symm) :
+  is_adj_matrix A.compl :=
+{ symm := by simp [h] }
+
+namespace is_adj_matrix
+
+variable {A}
+
+@[simp]
+lemma compl [has_zero α] [has_one α] (h : is_adj_matrix A) :
+  is_adj_matrix A.compl :=
+is_adj_matrix_compl A h.symm
+
+lemma compl_to_graph_eq [mul_zero_one_class α] [nontrivial α] (h : is_adj_matrix A) :
+  h.compl.to_graph = (h.to_graph)ᶜ :=
+begin
+  ext v w,
+  dsimp [simple_graph.has_compl, to_graph],
+  simp only [and_true, if_false_right_eq_and, matrix.compl],
+  split,
+  { intros h₁,
+    by_cases h₂ : v = w,
+    { simp * at * },
+    refine ⟨h₂, _⟩,
+    intros g,
+    rw [← apply_ne_zero_iff h] at g,
+    simp * at *  },
+  { rintros ⟨h₁, h₂⟩,
+    rw [apply_ne_one_iff h] at h₂,
+    simp [h₁, h₂] }
+end
+
+end is_adj_matrix
+
+end compl
+
+end matrix
+
+open matrix
 
 namespace simple_graph
 
-variables (G : simple_graph α) (R) [decidable_rel G.adj]
+variables (G : simple_graph V) [decidable_rel G.adj]
+variables (α)
 
-/-- `adj_matrix G R` is the matrix `A` such that `A i j = (1 : R)` if `i` and `j` are
+/-- `adj_matrix G α` is the matrix `A` such that `A i j = (1 : α)` if `i` and `j` are
   adjacent in the simple graph `G`, and otherwise `A i j = 0`. -/
-def adj_matrix : matrix α α R
+def adj_matrix [has_zero α] [has_one α] : matrix V V α
 | i j := if (G.adj i j) then 1 else 0
 
-variable {R}
+variable {α}
 
 @[simp]
-lemma adj_matrix_apply (v w : α) : G.adj_matrix R v w = if (G.adj v w) then 1 else 0 := rfl
+lemma adj_matrix_apply (v w : V) [has_zero α] [has_one α] :
+  G.adj_matrix α v w = if (G.adj v w) then 1 else 0 := rfl
 
 @[simp]
-theorem transpose_adj_matrix : (G.adj_matrix R)ᵀ = G.adj_matrix R :=
+theorem transpose_adj_matrix [has_zero α] [has_one α] :
+  (G.adj_matrix α)ᵀ = G.adj_matrix α :=
 by { ext, simp [adj_comm] }
 
-variables [fintype α]
+@[simp]
+lemma is_symm_adj_matrix [has_zero α] [has_one α] :
+  (G.adj_matrix α).is_symm :=
+transpose_adj_matrix G
+
+/-- The adjacency matrix of `G` is an adjacency matrix. -/
+@[simps]
+lemma is_adj_matrix_adj_matrix (α) [has_zero α] [has_one α] :
+  (G.adj_matrix α).is_adj_matrix :=
+{ zero_or_one := λ i j, by by_cases G.adj i j; simp [h] }
+
```

Comment:
Showing to_graph is a right inverse:
```lean
lemma to_graph_adj_matrix_eq [mul_zero_one_class α] [nontrivial α] :
  (is_adj_matrix_adj_matrix G α).to_graph = G :=
begin
  ext,
  simp only [is_adj_matrix.to_graph_adj, adj_matrix_apply, ite_eq_left_iff, zero_ne_one],
  apply not_not,
end
```

Reply:
Should I use `to_graph_adj_matrix_eq` or `adj_matrix_to_graph_eq`?

Reply:
You need both to show it's a two-sided inverse.

Reply:
I mean the name of the identifier for this one @kmill 
Which one makes more sense?

Reply:
Sorry, I'm not sure about what the right naming convention is here.  Use your judgment for now.

Reply:
@eric-wieser Eric, any suggestions?